### PR TITLE
Add subcodes to Validation core service

### PR DESCRIPTION
**Pull request by Yoann-Abbes**
:warning: depends on #1356 

## What does this PR do ?

Jira : [Subtask](https://jira.kaliop.net/browse/KZL-1198) from this [story](https://jira.kaliop.net/browse/KZL-1192)

Add subcodes to Validation core service
adapt unit test.

### How should this be manually tested?

unit/functional tests

### Boyscout

80 columns

depends on https://github.com/kuzzleio/kuzzle/pull/1356

### CODE REVIEW

File: lib/api/core/validation/types/date.js [hunk 1]
```diff
@@ -221,11 +227,11 @@ function convertRangeValue(name, value) {
   } else if (typeof value === 'number') { // epoch or epoch-millis
     converted = moment.utc(value);
   } else {
-    throw new PreconditionError(`Option "range.${name}": invalid format`);
+    errorsManager.throw('internal', 'validation', 'invalid_range_format', `range.${name}`);
   }
 
   if (!converted.isValid()) {
```
Comment (kuzzle) [inline lib/api/core/validation/types/date.js]: MAJOR TypeError can be thrown as "converted" might be null or undefined here. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS2259)

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: lib/api/core/validation/types/date.js [hunk 2]
```diff
@@ -165,38 +170,53 @@ class DateType extends BaseType {
    * @throws {PreconditionError}
    */
   validateFieldSpecification(typeOptions) {
+<<<<<<< HEAD
```
Comment (kuzzle) [inline lib/api/core/validation/types/date.js]: MAJOR Parse error at line 173 column 2:

163: 
164:     return true;
165:   }
166: 
167:   /**
168:    * @param {TypeOptions} typeOptions
169:    * @return {TypeOptions}
170:    * @throws {PreconditionError}
171:    */
172:   validateFieldSpecification(typeOptions) {
173: <<<<<<< HEAD
      ^
174:     if (typeOptions.hasOwnProperty('formats')) {
175:       if (!Array.isArray(typeOptions.formats) || typeOptions.formats.length === 0) {
176:         errorsManager.throw('internal', 'validation', 'invalid_formats_option');
177: =======
178:     if (_.has(typeOptions, 'formats')) {
179:       if (
180:         !Array.isArray(typeOptions.formats)
181:         || typeOptions.formats.length === 0
182:       ) {
183:         throw new PreconditionError('Invalid "formats" option: must be a non-empty array'); [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AParsingError)

File: lib/api/core/validation/types/numeric.js [hunk 3]
```diff
@@ -67,19 +74,36 @@ class NumericType extends BaseType {
...
   validateFieldSpecification(typeOptions) {
-    if (typeOptions.hasOwnProperty('range')) {
+    if (Object.prototype.hasOwnProperty.call(typeOptions, 'range')) {
       if (!this.checkAllowedProperties(typeOptions.range, ['min', 'max'])) {
-        throw new PreconditionError('Invalid "range" option definition');
+        errorsManager.throw('internal', 'validation', 'invalid_range_option');
       }
 
       for (const prop of ['min', 'max']) {
+<<<<<<< HEAD
```
Comment (kuzzle) [inline lib/api/core/validation/types/numeric.js]: MAJOR Parse error at line 83 column 2:

73:    * @return {TypeOptions}
74:    * @throws {PreconditionError}
75:    */
76:   validateFieldSpecification(typeOptions) {
77:     if (Object.prototype.hasOwnProperty.call(typeOptions, 'range')) {
78:       if (!this.checkAllowedProperties(typeOptions.range, ['min', 'max'])) {
79:         errorsManager.throw('internal', 'validation', 'invalid_range_option');
80:       }
81: 
82:       for (const prop of ['min', 'max']) {
83: <<<<<<< HEAD
     ^
84:         if (typeOptions.range.hasOwnProperty(prop) && typeof typeOptions.range[prop] !== 'number') {
85:           errorsManager.throw('internal', 'validation', 'invalid_range_type', prop);
86:         }
87:       }
88: 
89:       if (typeOptions.range.hasOwnProperty('min') && typeOptions.range.hasOwnProperty('max') && typeOptions.range.max < typeOptions.range.min) {
90:         errorsManager.throw('internal', 'validation', 'invalid_range');
91: =======
92:         if (
93:           Object.prototype.hasOwnProperty.call(typeOptions.range, prop) [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AParsingError)

File: lib/api/core/validation/index.js [hunk 4]
```diff
@@ -187,17 +209,37 @@ class Validation {
...
+              specSubset.children);
           }
+        } else {
+          documentSubset[fieldName] = this.recurseApplyDefault(
+            isUpdate,
+            field,
+            specSubset.children
+          );
         }
         else {
```
Comment (kuzzle) [inline lib/api/core/validation/index.js]: MAJOR Parse error at line 230 column 9:

220:               field[i],
221:               specSubset.children);
222:           }
223:         } else {
224:           documentSubset[fieldName] = this.recurseApplyDefault(
225:             isUpdate,
226:             field,
227:             specSubset.children
228:           );
229:         }
230:         else {
             ^
231:           documentSubset[fieldName] = this.recurseApplyDefault(
232:             isUpdate,
233:             field,
234:             specSubset.children);
235:         }
236:       }
237:       else if (
238:         specSubset.defaultValue
239:         && (
240:           field === null [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AParsingError)

**Comment by kuzzle** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 3 major
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L822](https://github.com/kuzzleio/kuzzle/blob/cd6abbdd707eb9254559c2bbb1a74c86a6a0c94f/lib/api/core/plugins/pluginsManager.js#L822): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by kuzzle** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L822](https://github.com/kuzzleio/kuzzle/blob/2302b8bfc6edc8ba02ea135a72cb60d76ba3e7d6/lib/api/core/plugins/pluginsManager.js#L822): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L822](https://github.com/kuzzleio/kuzzle/blob/b4d79d68599708e6c5607fa7a90af2a0707f4965/lib/api/core/plugins/pluginsManager.js#L822): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L822](https://github.com/kuzzleio/kuzzle/blob/a72cdd9d798429c9c13d07a952cf94013b87041f/lib/api/core/plugins/pluginsManager.js#L822): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L822](https://github.com/kuzzleio/kuzzle/blob/418f29f0caba129e80622c0ed298f3f3398592f5/lib/api/core/plugins/pluginsManager.js#L822): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: lib/api/core/validation/types/geoShape.js [hunk 5]
```diff
@@ -230,7 +259,11 @@ function isPoint (point) {
     return false;
   }
 
-  return !(point[0] < -180 || point[0] > 180 || point[1] < -90 || point[1] > 90);
+  return !(point[0] < -180
```
Comment (alexandrebouthinon) [inline lib/api/core/validation/types/geoShape.js]:
```suggestion
  return !(
    point[0] < -180
```

File: package.json [hunk 6]
```diff
@@ -32,30 +32,30 @@
     "@babel/plugin-transform-runtime": "^7.4.4",
```
Comment (alexandrebouthinon) [inline package.json]: Are the `package.json` changes wanted?

Comment (Yoann-Abbes) [reply package.json]: Yes since i have merged https://github.com/kuzzleio/kuzzle/pull/1356 into this pr

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L822](https://github.com/kuzzleio/kuzzle/blob/ab263557770333c44892ae42b876ae73785acfd3/lib/api/core/plugins/pluginsManager.js#L822): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: lib/api/core/validation/types/enum.js [hunk 7]
```diff
@@ -47,7 +47,9 @@ class EnumType extends BaseType {
     }
 
     if (typeOptions.values.indexOf(fieldValue) === -1) {
-      errorMessages.push(`The field only accepts following values: "${typeOptions.values.join(', ')}".`);
+      errorMessages.push(
+        `The field only accepts following values: "${typeOptions.values.join(', ')}".`
+      );
```
Comment (Aschen) [inline lib/api/core/validation/types/enum.js]:
```suggestion
```

File: lib/api/core/validation/types/enum.js [hunk 8]
```diff
@@ -47,7 +47,9 @@ class EnumType extends BaseType {
     }
 
     if (typeOptions.values.indexOf(fieldValue) === -1) {
-      errorMessages.push(`The field only accepts following values: "${typeOptions.values.join(', ')}".`);
+      errorMessages.push(
+        `The field only accepts following values: "${typeOptions.values.join(', ')}".`
```
Comment (Aschen) [inline lib/api/core/validation/types/enum.js]:
```suggestion
        `The field only accepts following values: "${typeOptions.values.join(', ')}".`);
```

File: lib/api/core/validation/types/enum.js [hunk 9]
```diff
@@ -60,18 +62,34 @@ class EnumType extends BaseType {
...
-    if (!Array.isArray(typeOptions.values) || typeOptions.values.length === 0) {
-      throw new PreconditionError('Option "values" must be a non-empty array');
+    if (
+      !Array.isArray(typeOptions.values)
+      || typeOptions.values.length === 0
+    ) {
+      errorsManager.throw(
+        'internal',
+        'validation',
+        'values_option_cannot_be_empty'
```
Comment (Aschen) [inline lib/api/core/validation/types/enum.js]: The last parenthesis should be on the last line.
There is other occurrences of this format mistake

File: lib/api/core/validation/types/enum.js [hunk 10]
```diff
@@ -47,7 +47,9 @@ class EnumType extends BaseType {
     }
 
     if (typeOptions.values.indexOf(fieldValue) === -1) {
```
Comment (Aschen) [inline lib/api/core/validation/types/enum.js]:
```suggestion
    if (!typeOptions.values.includes(fieldValue)) {
```

File: lib/api/core/validation/types/geoShape.js [hunk 11]
```diff
@@ -77,7 +81,10 @@ class GeoShapeType extends BaseType {
       case 'polygon':
       case 'multipolygon':
         coordinateValidation = isPolygon;
-        if (shape.orientation && allowedOrientations.indexOf(shape.orientation) === -1) {
+        if (
+          shape.orientation
+          && allowedOrientations.indexOf(shape.orientation) === -1
```
Comment (Aschen) [inline lib/api/core/validation/types/geoShape.js]:
```suggestion
          && !allowedOrientations.includes(shape.orientation)
```

File: lib/api/core/validation/types/geoShape.js [hunk 12]
```diff
@@ -202,15 +227,19 @@ class GeoShapeType extends BaseType {
...
+      if (
+        !Array.isArray(typeOptions.shapeTypes)
+        || typeOptions.shapeTypes.length === 0
+      ) {
+        errorsManager.throw('internal', 'validation', 'shapetypes_option_cannot_be_empty');
       }
 
-      const invalid = typeOptions.shapeTypes.filter(shape => allowedShapeTypes.indexOf(shape) === -1);
+      const invalid = typeOptions.shapeTypes.filter(
+        shape => allowedShapeTypes.indexOf(shape) === -1);
```
Comment (Aschen) [inline lib/api/core/validation/types/geoShape.js]:
```suggestion
        shape => !allowedShapeTypes.includes(shape));
```

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L822](https://github.com/kuzzleio/kuzzle/blob/f7020272d22c2decaafd878126a2565d4b9b849a/lib/api/core/plugins/pluginsManager.js#L822): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L822](https://github.com/kuzzleio/kuzzle/blob/e8d1d30ba8268857649c34a51bf06e1545341c13/lib/api/core/plugins/pluginsManager.js#L822): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L822](https://github.com/kuzzleio/kuzzle/blob/3f27e77def18550abb371a8714bcd734304003b0/lib/api/core/plugins/pluginsManager.js#L822): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L822](https://github.com/kuzzleio/kuzzle/blob/1ab37cb41a28549db31b3751fa1e1f1111de7732/lib/api/core/plugins/pluginsManager.js#L822): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: lib/api/core/validation/baseType.js [hunk 13]
```diff
@@ -52,11 +52,15 @@ class BaseType {
    * @returns {boolean}
    */
   checkAllowedProperties(object, allowedProperties) {
-    if (typeof object !== 'object' || Array.isArray(object) || object === null) {
+    if (typeof object !== 'object'
```
Comment (Njuelle) [inline lib/api/core/validation/baseType.js]:
```suggestion
    if ( typeof object !== 'object'
```

File: lib/api/core/validation/types/date.js [hunk 14]
```diff
@@ -171,16 +171,23 @@ class DateType extends BaseType {
    */
   validateFieldSpecification(typeOptions) {
     if (_.has(typeOptions, 'formats')) {
-      if (
-        !Array.isArray(typeOptions.formats)
+      if (!Array.isArray(typeOptions.formats)
```
Comment (Njuelle) [inline lib/api/core/validation/types/date.js]:
```suggestion
      if ( !Array.isArray(typeOptions.formats)
```

File: lib/api/core/validation/types/enum.js [hunk 15]
```diff
@@ -46,8 +46,9 @@ class EnumType extends BaseType {
       return false;
     }
 
-    if (typeOptions.values.indexOf(fieldValue) === -1) {
-      errorMessages.push(`The field only accepts following values: "${typeOptions.values.join(', ')}".`);
+    if (!typeOptions.values.includes(fieldValue)) {
```
Comment (Njuelle) [inline lib/api/core/validation/types/enum.js]:
```suggestion
    if ( !typeOptions.values.includes(fieldValue)) {
```

File: lib/api/core/validation/types/enum.js [hunk 16]
```diff
@@ -61,17 +62,30 @@ class EnumType extends BaseType {
...
   validateFieldSpecification(typeOptions) {
     if (!Object.prototype.hasOwnProperty.call(typeOptions, 'values')) {
-      throw new PreconditionError('Option "values" is required');
+      errorsManager.throw('internal', 'validation', 'missing_values_option');
+
     }
 
-    if (!Array.isArray(typeOptions.values) || typeOptions.values.length === 0) {
-      throw new PreconditionError('Option "values" must be a non-empty array');
+    if (!Array.isArray(typeOptions.values)
```
Comment (Njuelle) [inline lib/api/core/validation/types/enum.js]:
```suggestion
    if ( !Array.isArray(typeOptions.values)
```

File: lib/api/core/validation/types/geoShape.js [hunk 17]
```diff
@@ -77,7 +81,9 @@ class GeoShapeType extends BaseType {
       case 'polygon':
       case 'multipolygon':
         coordinateValidation = isPolygon;
-        if (shape.orientation && allowedOrientations.indexOf(shape.orientation) === -1) {
+        if (shape.orientation
```
Comment (Njuelle) [inline lib/api/core/validation/types/geoShape.js]:
```suggestion
        if ( shape.orientation
```

File: lib/api/core/validation/types/geoShape.js [hunk 18]
```diff
@@ -163,7 +174,9 @@ class GeoShapeType extends BaseType {
       result = false;
     }
 
-    if (shape.type !== 'geometrycollection' && (!shape.coordinates || !Array.isArray(shape.coordinates))) {
+    if (shape.type !== 'geometrycollection'
```
Comment (Njuelle) [inline lib/api/core/validation/types/geoShape.js]:
```suggestion
    if ( shape.type !== 'geometrycollection'
```

File: lib/api/core/validation/types/geoShape.js [hunk 19]
```diff
@@ -178,7 +191,10 @@ class GeoShapeType extends BaseType {
       result = false;
     }
 
-    if (shape.type !== 'polygon' && shape.type !== 'multipolygon' && shape.orientation) {
+    if (shape.type !== 'polygon'
```
Comment (Njuelle) [inline lib/api/core/validation/types/geoShape.js]:
```suggestion
    if ( shape.type !== 'polygon'
```

File: lib/api/core/validation/types/geoShape.js [hunk 20]
```diff
@@ -188,7 +204,9 @@ class GeoShapeType extends BaseType {
       result = false;
     }
 
-    if (shape.type === 'geometrycollection' && (!shape.geometries || !Array.isArray(shape.geometries))) {
+    if (shape.type === 'geometrycollection'
```
Comment (Njuelle) [inline lib/api/core/validation/types/geoShape.js]:
```suggestion
    if ( shape.type === 'geometrycollection'
```

File: lib/api/core/validation/types/numeric.js [hunk 21]
```diff
@@ -75,24 +73,32 @@ class NumericType extends BaseType {
...
+        errorsManager.throw(
+          'internal',
+          'validation',
+          'invalid_range_option');
       }
 
       for (const prop of ['min', 'max']) {
-        if (
-          Object.prototype.hasOwnProperty.call(typeOptions.range, prop)
+        if (Object.prototype.hasOwnProperty.call(typeOptions.range, prop)
```
Comment (Njuelle) [inline lib/api/core/validation/types/numeric.js]:
```suggestion
        if ( Object.prototype.hasOwnProperty.call(typeOptions.range, prop)
```

File: lib/api/core/validation/types/numeric.js [hunk 22]
```diff
@@ -75,24 +73,32 @@ class NumericType extends BaseType {
...
+            'internal',
+            'validation',
+            'invalid_range_type',
+            prop);
         }
       }
 
-      if (
-        Object.prototype.hasOwnProperty.call(typeOptions.range, 'min')
+      if (Object.prototype.hasOwnProperty.call(typeOptions.range, 'min')
```
Comment (Njuelle) [inline lib/api/core/validation/types/numeric.js]:
```suggestion
      if ( Object.prototype.hasOwnProperty.call(typeOptions.range, 'min')
```

File: lib/api/core/validation/types/object.js [hunk 23]
```diff
@@ -41,7 +41,10 @@ class ObjectType extends BaseType {
    * @returns {boolean}
    */
   validate(typeOptions, fieldValue, errorMessages) {
-    if (fieldValue === null || typeof fieldValue !== 'object' || Array.isArray(fieldValue)) {
+    if (fieldValue === null
```
Comment (Njuelle) [inline lib/api/core/validation/types/object.js]:
```suggestion
    if ( fieldValue === null
```

File: lib/api/core/validation/types/object.js [hunk 24]
```diff
@@ -55,11 +58,10 @@ class ObjectType extends BaseType {
    * @throws {PreconditionError}
    */
   validateFieldSpecification(typeOptions) {
-    if (
-      Object.prototype.hasOwnProperty.call(typeOptions, 'strict')
+    if (Object.prototype.hasOwnProperty.call(typeOptions, 'strict')
```
Comment (Njuelle) [inline lib/api/core/validation/types/object.js]:
```suggestion
    if ( Object.prototype.hasOwnProperty.call(typeOptions, 'strict')
```

File: lib/api/core/validation/types/string.js [hunk 25]
```diff
@@ -50,16 +50,14 @@ class StringType extends BaseType {
       return true;
     }
 
-    if (
-      Object.prototype.hasOwnProperty.call(typeOptions.length, 'min')
+    if (Object.prototype.hasOwnProperty.call(typeOptions.length, 'min')
```
Comment (Njuelle) [inline lib/api/core/validation/types/string.js]:
```suggestion
    if ( Object.prototype.hasOwnProperty.call(typeOptions.length, 'min')
```

File: lib/api/core/validation/types/string.js [hunk 26]
```diff
@@ -50,16 +50,14 @@ class StringType extends BaseType {
...
       && fieldValue.length < typeOptions.length.min
     ) {
       errorMessages.push(`Invalid string length. Expected min: ${typeOptions.length.min}. Received: ${fieldValue.length} ("${fieldValue}")`);
       return false;
     }
 
-    if (
-      Object.prototype.hasOwnProperty.call(typeOptions.length, 'max')
+    if (Object.prototype.hasOwnProperty.call(typeOptions.length, 'max')
```
Comment (Njuelle) [inline lib/api/core/validation/types/string.js]:
```suggestion
    if ( Object.prototype.hasOwnProperty.call(typeOptions.length, 'max')
```

File: lib/api/core/validation/types/string.js [hunk 27]
```diff
@@ -77,24 +75,27 @@ class StringType extends BaseType {
...
     if (Object.prototype.hasOwnProperty.call(typeOptions, 'length')) {
       if (!this.checkAllowedProperties(typeOptions.length, ['min', 'max'])) {
-        throw new PreconditionError('Invalid "length" option definition');
+        errorsManager.throw('internal', 'validation', 'invalid_length_option');
       }
 
       for (const prop of ['min', 'max']) {
-        if (
-          Object.prototype.hasOwnProperty.call(typeOptions.length, prop)
+        if (Object.prototype.hasOwnProperty.call(typeOptions.length, prop)
```
Comment (Njuelle) [inline lib/api/core/validation/types/string.js]:
```suggestion
        if ( Object.prototype.hasOwnProperty.call(typeOptions.length, prop)
```

File: lib/api/core/validation/types/string.js [hunk 28]
```diff
@@ -77,24 +75,27 @@ class StringType extends BaseType {
...
+            'validation',
+            'invalid_length_type',
+            prop
+          );
         }
       }
 
-      if (
-        Object.prototype.hasOwnProperty.call(typeOptions.length, 'min')
+      if (Object.prototype.hasOwnProperty.call(typeOptions.length, 'min')
```
Comment (Njuelle) [inline lib/api/core/validation/types/string.js]:
```suggestion
      if ( Object.prototype.hasOwnProperty.call(typeOptions.length, 'min')
```

File: lib/api/core/validation/types/enum.js [hunk 29]
```diff
@@ -46,8 +46,9 @@ class EnumType extends BaseType {
       return false;
     }
 
-    if (typeOptions.values.indexOf(fieldValue) === -1) {
-      errorMessages.push(`The field only accepts following values: "${typeOptions.values.join(', ')}".`);
+    if ( !typeOptions.values.includes(fieldValue)) {
```
Comment (Njuelle) [inline lib/api/core/validation/types/enum.js]:
```suggestion
    if (!typeOptions.values.includes(fieldValue)) {
```

**Pull request merged**